BUG: Add home directory to path for global .numpy-site.cfg file.

bpbrown opened the pull request:
Bug fix for issue #11581

Commit:
Adds home directory to pathing for global .numpy-site.cfg file.

bpbrown commented:
Issue: current commit message does not follow numpy contribution guidelines.  Sorry.  What is standard practice in the numpy community in revising commit messages?

charris commented:
You can edit the commit and force push, changing the history is OK when the PR is being worked on.

rgommers commented on numpy/distutils/system_info.py:
```diff
@@ -540,7 +540,7 @@ def __init__(self,
         self.cp = ConfigParser(defaults)
         self.files = []
-        self.files.extend(get_standard_file('.numpy-site.cfg'))
+        self.files.extend(get_standard_file('~/.numpy-site.cfg'))
```
Interesting that we missed this for so long. 

This change is not cross-platform, should be `os.path.join(os.environ['HOME'], 'numpy-site.cfg')`

pv commented:
Is this change correct? The `get_standard_file` function states it
searches also HOME.

bpbrown commented:
It does not in practice, in my testing at least.

Maybe this is a get_standard_file bug.

On Tue, Jul 17, 2018 at 10:16 AM Pauli Virtanen <notifications@github.com>
wrote:

> Is this change correct? The `get_standard_file` function states it
> searches also HOME.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/11582#issuecomment-405640235>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFvw4WbTiBwdfOY8L9aDowwABGUpdosJks5uHg3DgaJpZM4VS3Jz>
> .
>

rgommers commented:
Works for me:
```
In [1]: from numpy.distutils.system_info import get_standard_file

In [2]: get_standard_file('numpy-site.cfg')
Out[2]: ['/home/rgommers/numpy-site.cfg']
```
The `get_standard_file` code looks correct. @bpbrown could you try to figure out why on your system `os.path.expanduser` (https://docs.python.org/3/library/os.path.html#os.path.expanduser) doesn't do what you expect?

bpbrown commented:
Ralf,
     Thanks. I’ll test and get back with a response.

—Ben

On Sat, Jul 21, 2018 at 2:40 PM Ralf Gommers <notifications@github.com>
wrote:

> Works for me:
>
> In [1]: from numpy.distutils.system_info import get_standard_file
>
> In [2]: get_standard_file('numpy-site.cfg')
> Out[2]: ['/home/rgommers/numpy-site.cfg']
>
> The get_standard_file code looks correct. @bpbrown
> <https://github.com/bpbrown> could you try to figure out why on your
> system os.path.expanduser (
> https://docs.python.org/3/library/os.path.html#os.path.expanduser)
> doesn't do what you expect?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/11582#issuecomment-406825078>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFvw4axFlka8-w-nAjZSClUyijbx2ohMks5uI5-1gaJpZM4VS3Jz>
> .
>

bpbrown commented:
This is interesting.  get_standard_file() is correctly expanding my home
path:
```
>>> from numpy.distutils.system_info import get_standard_file
>>> get_standard_file('.numpy-site.cfg')
['/home5/bpbrown/.numpy-site.cfg', '/home5/bpbrown/.numpy-site.cfg']
>>>
```

But pip installed numpy seems to not follow the commands in the global
~/.numpy-site.cfg.  That file is:
```
bpbrown@pfe25:~> more ~/.numpy-site.cfg
[mkl]
library_dirs =
/nasa/intel/Compiler/2018.0.128/compilers_and_libraries_2018.0.128/linux/mkl/lib/intel64/
include_dirs =
/nasa/intel/Compiler/2018.0.128/compilers_and_libraries_2018.0.128/linux/mkl/include
mkl_libs = mkl_rt
lapack_libs =
bpbrown@pfe25:~>
```
and is trying to link against MKL BLAS.

Post-install, numpy is not linked against MKL:
```
>>> print(numpy.__config__.show())
blas_mkl_info:
  NOT AVAILABLE
<...>
```

whereas the expected behviour (pip installing using proposed git commit
ba210fd211dda054e351c74b50ace4dd76dd33ab) is:
```
>>> print(numpy.__config__.show())
blas_mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs =
['/nasa/intel/Compiler/2018.0.128/compilers_and_libraries_2018.0.128/linux/mkl/lib/intel64']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs =
['/nasa/intel/Compiler/2018.0.128/compilers_and_libraries_2018.0.128/linux/mkl',
'/nasa/intel/Compiler/2018.0.128/compilers_and_libraries_2018.0.128/linux/mkl/include',
'/nasa/intel/Compiler/2018.0.128/compilers_and_libraries_2018.0.128/linux/mkl/lib']
```
and we're correctly showing links against the MKL BLAS.

Any idea what's going wrong/what I'm doing wrong/why the entries of
~/.numpy-site.cfg are being disregarded in the vanilla pip install?


Thanks,
--Ben


On Sat, Jul 21, 2018 at 5:00 PM, Benjamin Brown <bpbrown@colorado.edu>
wrote:

> Ralf,
>      Thanks. I’ll test and get back with a response.
>
> —Ben
>
> On Sat, Jul 21, 2018 at 2:40 PM Ralf Gommers <notifications@github.com>
> wrote:
>
>> Works for me:
>>
>> In [1]: from numpy.distutils.system_info import get_standard_file
>>
>> In [2]: get_standard_file('numpy-site.cfg')
>> Out[2]: ['/home/rgommers/numpy-site.cfg']
>>
>> The get_standard_file code looks correct. @bpbrown
>> <https://github.com/bpbrown> could you try to figure out why on your
>> system os.path.expanduser (https://docs.python.org/3/
>> library/os.path.html#os.path.expanduser) doesn't do what you expect?
>>
>> —
>> You are receiving this because you were mentioned.
>>
>>
>> Reply to this email directly, view it on GitHub
>> <https://github.com/numpy/numpy/pull/11582#issuecomment-406825078>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/AFvw4axFlka8-w-nAjZSClUyijbx2ohMks5uI5-1gaJpZM4VS3Jz>
>> .
>>
>

rgommers commented:
> But pip installed numpy seems to not follow the commands in the global ~/.numpy-site.cfg.

How exactly did you invoke pip, and can you provide the verbose output of the pip install (e.g. `pip install . -v`)?

bpbrown commented:
The pip install line being used is:

    pip3 install numpy

So nothing special.  The output from

    pip3 install numpy -v >  numpy_install.log

is
```
(dedalus)bpbrown@pfe21:~> more numpy_install.log 
Created temporary directory: /tmp/pip-ephem-wheel-cache-ctjlcfca
Created temporary directory: /tmp/pip-req-tracker-cn35kr51
Created requirements tracker '/tmp/pip-req-tracker-cn35kr51'
Created temporary directory: /tmp/pip-install-14c1l1fy
Collecting numpy
  1 location(s) to search for versions of numpy:
  * https://pypi.org/simple/numpy/
  Getting page https://pypi.org/simple/numpy/
  Looking up "https://pypi.org/simple/numpy/" in the cache
  Current age based on date: 13
  Freshness lifetime from max-age: 600
  Freshness lifetime from request max-age: 600
  The response is "fresh", returning cached response
  600 > 13
  Analyzing links from page https://pypi.org/simple/numpy/

<... bunch of link tests/skips based on installs that aren't compatible ...>
Found link https://files.pythonhosted.org/packages/3a/20/c81632328b1a4e1db65f45c0a1350a9c5341fd4bbb8ea66cdd98da56fe2e/numpy-1.15.0.zip#sha256=f28e73cf18d37a413f7d5de35d024e6b98f14566a10d82100f9dc491a7d449f9 (from https://pypi.org/simple/numpy/) (requires-python:>=2.7,!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*), version: 1.15.0
  Using version 1.15.0 (newest of versions: 1.3.0, 1.4.1, 1.5.0, 1.5.1, 1.6.0, 1.6.1, 1.6.2, 1.7.0, 1.7.1, 1.7.2, 1.8.0, 1.8.1, 1.8.2, 1.9.0, 1.9.1, 1.9.2, 1.9.3, 1.10.0.post2, 1.10.1, 1.10.2, 1.10.4, 1.11.0, 1.11.1, 1.11.2, 1.11.3, 1.12.0, 1.12.1, 1.13.0, 1.13.1, 1.13.3, 1.14.0, 1.14.1, 1.14.2, 1.14.3, 1.14.4, 1.14.5, 1.15.0)
  Created temporary directory: /tmp/pip-unpack-bxjafq7l
  Looking up "https://files.pythonhosted.org/packages/27/92/c01d3a6c58ceab0e6ec36ad3af41bc076014cc916afcb979ab4c9558f347/numpy-1.15.0-cp37-cp37m-manylinux1_x86_64.whl" in the cache
  Current age based on date: 18265
  Ignoring unknown cache-control directive: immutable
  Freshness lifetime from max-age: 365000000
  The response is "fresh", returning cached response
  365000000 > 18265
  Using cached https://files.pythonhosted.org/packages/27/92/c01d3a6c58ceab0e6ec36ad3af41bc076014cc916afcb979ab4c9558f347/numpy-1.15.0-cp37-cp37m-manylinux1_x86_64.whl
  Downloading from URL https://files.pythonhosted.org/packages/27/92/c01d3a6c58ceab0e6ec36ad3af41bc076014cc916afcb979ab4c9558f347/numpy-1.15.0-cp37-cp37m-manylinux1_x86_64.whl#sha256=50718eea8e77a1bedcc85befd22c8dbf5a24c9d2c0c1e36bbb8d7a38da847eb3 (from https://pypi.org/simple/numpy/) (requires-python:>=2.7,!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*)
  Added numpy from https://files.pythonhosted.org/packages/27/92/c01d3a6c58ceab0e6ec36ad3af41bc076014cc916afcb979ab4c9558f347/numpy-1.15.0-cp37-cp37m-manylinux1_x86_64.whl#sha256=50718eea8e77a1bedcc85befd22c8dbf5a24c9d2c0c1e36bbb8d7a38da847eb3 to build tracker '/tmp/pip-req-tracker-cn35kr51'
  Removed numpy from https://files.pythonhosted.org/packages/27/92/c01d3a6c58ceab0e6ec36ad3af41bc076014cc916afcb979ab4c9558f347/numpy-1.15.0-cp37-cp37m-manylinux1_x86_64.whl#sha256=50718eea8e77a1bedcc85befd22c8dbf5a24c9d2c0c1e36bbb8d7a38da847eb3 from build tracker '/tmp/pip-req-tracker-cn35kr51'
Installing collected packages: numpy

  changing mode of /home5/bpbrown/dedalus/bin/conv-template to 755
  changing mode of /home5/bpbrown/dedalus/bin/f2py to 755
  changing mode of /home5/bpbrown/dedalus/bin/from-template to 755
Successfully installed numpy-1.15.0
Cleaning up...
Removed build tracker '/tmp/pip-req-tracker-cn35kr51'
(dedalus)bpbrown@pfe21:~> 
```
No obvious error flags about site.cfg, and we're still not finding the specified MKL libraries:
```
>>> import numpy
>>> print(numpy.__config__.show())
blas_mkl_info:
  NOT AVAILABLE
<... more entries ...>
```
 Let me know what to test next to clear this up, and thank you for your continued help in tracking down where this problem lies.

rgommers commented:
You're just downloading a binary wheel there, not building numpy from source. So there's no linking against any BLAS to be done at that stage. If you want a numpy linked against MKL, either install it from with `conda` from the defaults channel, or actually build from source.

bpbrown commented:
Thank you.  Building from source worked.  Could we please add this to the numpy documentation?
